File: year_2021/solutions/python/day_02.py
from typing import List
from pathlib import Path
from year_2021.solutions.python.utils.files import read_lines, INPUTS_FOLDER


def move_with_simple_commands(instructions_list: List[str]) -> int:
    position, depth = 0, 0
    for instruction in instructions_list:
        match instruction.split():
            case ['forward', value]:
                position += int(value)
            case ['down', value]:
                depth += int(value)
            case ['up', value]:
                depth -= int(value)
    return position * depth


def move_with_advanced_commands(instructions_list: List[str]) -> int:
    position, depth, aim = 0, 0, 0
    for instruction in instructions_list:
        match instruction.split():
            case ['forward', value]:
                position += int(value)
                depth += aim * int(value)
            case ['down', value]:
                aim += int(value)
            case ['up', value]:
                aim -= int(value)
    return position * depth


if __name__ == '__main__':
    input_file_path = INPUTS_FOLDER / Path(__file__).with_suffix('').name / 'input.txt'
    instructions: List[str] = read_lines(input_file_path=input_file_path, line_type=str)

    # Part 1
    part_1_result: int = move_with_simple_commands(instructions)
    assert part_1_result == 1459206
    print('Part 1 result :', part_1_result)

    # Part 2
    part_2_result: int = move_with_advanced_commands(instructions)
    assert part_2_result == 1320534480
    print('Part 2 result :', part_2_result)

File: year_2021/solutions/julia/day_02.jl
include(joinpath(@__DIR__(), "utils", "files.jl"))


function move_with_simple_commands(instructions_list::Array{String})::Int
    position, depth = 0, 0
    for instruction in instructions_list
        command, value = split(instruction)
        if command == "forward"
            position += parse(Int, value)
        elseif command == "down"
            depth += parse(Int, value)
        elseif command == "up"
            depth -= parse(Int, value)
        end
    end
    return position * depth
end


function move_with_advanced_commands(instructions_list::Array{String})::Int
    position, depth, aim = 0, 0, 0
    for instruction in instructions_list
        command, value = split(instruction)
        if command == "forward"
            position += parse(Int, value)
            depth += aim * parse(Int, value)
        elseif command == "down"
            aim += parse(Int, value)
        elseif command == "up"
            aim -= parse(Int, value)
        end
    end
    return position * depth
end


function main()
    input_file_path::String = joinpath(INPUTS_FOLDER, splitext(basename(@__FILE__))[1], "input.txt")
    instructions::Array{String} = readlines(input_file_path)
    
    # Part 1
    part_1_result::Int = move_with_simple_commands(instructions)
    @assert part_1_result == 1459206
    println("Part 1 result : ", part_1_result)

    # Part 2
    part_2_result::Int = move_with_advanced_commands(instructions)
    @assert part_2_result == 1320534480
    println("Part 2 result : ", part_2_result)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
